File: packages/catalog-python/src/ai37_a2ui_catalog/models/flex_table.py
from __future__ import annotations

from typing import Literal

from pydantic import Field, model_validator

from .shared import StrictModel, TextAlign


class FlexTableCell(StrictModel):
    content: str = Field(min_length=1)
    align: TextAlign = None
    colSpan: int = Field(default=None, ge=1)
    rowSpan: int = Field(default=None, ge=1)
    emphasis: Literal["normal", "muted", "strong"] = None


class FlexTableRow(StrictModel):
    cells: list[FlexTableCell] = Field(min_length=1)


def validate_header_grid(rows: list[FlexTableRow]) -> int:
    total_columns = sum(cell.colSpan or 1 for cell in rows[0].cells)
    active_spans = [0] * total_columns

    for row_index, row in enumerate(rows):
        if row_index > 0:
            active_spans = [max(0, span - 1) for span in active_spans]

        column_index = 0

        for cell in row.cells:
            while column_index < total_columns and active_spans[column_index] > 0:
                column_index += 1

            col_span = cell.colSpan or 1
            row_span = cell.rowSpan or 1

            if column_index + col_span > total_columns:
                raise ValueError(f"Header row {row_index} exceeds the available column span.")

            if any(active_spans[column_index + offset] > 0 for offset in range(col_span)):
                raise ValueError(f"Header row {row_index} overlaps an active rowSpan.")

            for offset in range(col_span):
                active_spans[column_index + offset] = row_span

            column_index += col_span

        if any(span == 0 for span in active_spans):
            raise ValueError(f"Header row {row_index} does not fully cover the table width.")

    return total_columns


class FlexTableProps(StrictModel):
    title: str = Field(default=None, min_length=1, max_length=120)
    caption: str = Field(default=None, min_length=1, max_length=240)
    dense: bool = None
    headerRows: list[FlexTableRow] = Field(min_length=1)
    bodyRows: list[FlexTableRow] = Field(min_length=1)

    @model_validator(mode="after")
    def validate_row_spans(self) -> "FlexTableProps":
        expected = validate_header_grid(self.headerRows)

        for index, row in enumerate(self.bodyRows):
            width = sum(cell.colSpan or 1 for cell in row.cells)
            if width != expected:
                raise ValueError(f"Body row {index} has inconsistent total span.")

        return self

File: packages/catalog-schemas/src/components/flex-table.ts
import {z} from 'zod';
import {type CatalogComponentDefinition, textAlignSchema} from '../types';

const emphasisSchema = z.enum(['normal', 'muted', 'strong']);

export const flexTableCellSchema = z
  .object({
    content: z.string().min(1),
    align: textAlignSchema.optional(),
    colSpan: z.number().int().min(1).optional(),
    rowSpan: z.number().int().min(1).optional(),
    emphasis: emphasisSchema.optional(),
  })
  .strict();

export const flexTableRowSchema = z
  .object({
    cells: z.array(flexTableCellSchema).min(1),
  })
  .strict();

const spanWidth = (row: z.infer<typeof flexTableRowSchema>) =>
  row.cells.reduce((sum, cell) => sum + (cell.colSpan ?? 1), 0);

function validateHeaderGrid(rows: z.infer<typeof flexTableRowSchema>[], ctx: z.RefinementCtx) {
  const totalColumns = spanWidth(rows[0]!);
  const activeSpans = Array.from({length: totalColumns}, () => 0);

  for (const [rowIndex, row] of rows.entries()) {
    if (rowIndex > 0) {
      for (let columnIndex = 0; columnIndex < activeSpans.length; columnIndex += 1) {
        activeSpans[columnIndex] = Math.max(0, activeSpans[columnIndex]! - 1);
      }
    }

    let columnIndex = 0;

    for (const [cellIndex, cell] of row.cells.entries()) {
      while (columnIndex < totalColumns && activeSpans[columnIndex]! > 0) {
        columnIndex += 1;
      }

      const colSpan = cell.colSpan ?? 1;
      const rowSpan = cell.rowSpan ?? 1;

      if (columnIndex + colSpan > totalColumns) {
        ctx.addIssue({
          code: z.ZodIssueCode.custom,
          message: `Header row ${rowIndex} exceeds the available column span.`,
          path: ['headerRows', rowIndex, 'cells', cellIndex],
        });
        return totalColumns;
      }

      for (let offset = 0; offset < colSpan; offset += 1) {
        if (activeSpans[columnIndex + offset]! > 0) {
          ctx.addIssue({
            code: z.ZodIssueCode.custom,
            message: `Header row ${rowIndex} overlaps an active rowSpan.`,
            path: ['headerRows', rowIndex, 'cells', cellIndex],
          });
          return totalColumns;
        }
      }

      for (let offset = 0; offset < colSpan; offset += 1) {
        activeSpans[columnIndex + offset] = rowSpan;
      }

      columnIndex += colSpan;
    }

    if (activeSpans.some(span => span === 0)) {
      ctx.addIssue({
        code: z.ZodIssueCode.custom,
        message: `Header row ${rowIndex} does not fully cover the table width.`,
        path: ['headerRows', rowIndex],
      });
      return totalColumns;
    }
  }

  return totalColumns;
}

export const flexTablePropsSchema = z
  .object({
    title: z.string().min(1).max(120).optional(),
    caption: z.string().min(1).max(240).optional(),
    dense: z.boolean().optional(),
    headerRows: z.array(flexTableRowSchema).min(1),
    bodyRows: z.array(flexTableRowSchema).min(1),
  })
  .strict()
  .superRefine((value, ctx) => {
    const headerWidth = validateHeaderGrid(value.headerRows, ctx);

    for (const [index, row] of value.bodyRows.entries()) {
      if (spanWidth(row) !== headerWidth) {
        ctx.addIssue({
          code: z.ZodIssueCode.custom,
          message: `Body row ${index} has inconsistent total span.`,
          path: ['bodyRows', index],
        });
      }
    }
  });

export type FlexTableProps = z.infer<typeof flexTablePropsSchema>;

export const flexTableDefinition: CatalogComponentDefinition<typeof flexTablePropsSchema> = {
  name: 'FlexTable',
  slug: 'flex-table',
  description:
    'A flexible table with multi-row headers and row or column spans. Use it for grouped metrics, matrix reports, or any dataset that requires composite headers.',
  schema: flexTablePropsSchema,
};
